### PR TITLE
Add tests for new APIs for chassis and convert existing API tests to run on chassis

### PR DESCRIPTION
**Pull request by rawal01**
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

#### What is the motivation for this PR?
Add new test cases for api  for test plan introduced by PR #2695 https://github.com/Azure/sonic-mgmt/pull/2695 (refer to sections 2 to 6)
Need to run existing api tests against a Sonice chassis but all of the tests were using duthost and hence will always run one of duthost of chassis i.e supervisor or line card. Converted these tests to use the enum_* fixture per hwsku for dut selection. Instead of running gather facts as fixture get facts within compare function for selected duthost.

  
#### How did you do it?

Following changes made to api helper files in folder sonic-mgmt/tests/common/helpers/platform_api/
    chassis.py:
	add new api calls: get_module_index, get_supervisor_slot, get_my_slot, is_modular_chassis
	
	fan_drawer.py:
	add new api call: get_maximum_consumed_power
	
	module.py:
	add new api calls: get_description,	get_slot, get_type,	get_oper_status, get_midplane_ip, 
	is_midplane_reachable, get_maximum_consumed_power, reboot, set_admin_state
	
	psu.py:
	add new api calls: get_maximum_supplied_power, set_status_master_led, get_status_master_led
	
	thermal.py:
	add new api calls: get_minimum_recorded, get_maximum_recorded
	
changes maded to api tests to support Sonic Chassis and add tests for new apis:
	
	conftest.py:
	changes to run tests on  Sonic chassis
	change  plaftorm_api_conn to support multidut by using duthosts and enum_rand_one_per_hwsku_hostname to select dut at function level, 
	change from getting ip from eth0 to duthost.mgp_ip to extract ip address for DUT 
	
	change start_platform_service  to support multidut by using duthosts and enum_rand_one_per_hwsku_hostname to select dut at function level,
	change from getting ip from eth0 to duthost.mgp_ip to extract ip address for DUT 
	
	change in stop_platform_api_service: to support on all duts where the service was started
	
	
for all test modules under api these are common changes :
	Remove gather_facts fixture and get facts within compare_value_with_platform_facts per duthost from test
	changes changes to run tests on  Sonic chassis, Replace duthost with duthosts and     enum_rand_one_per_hwsku_hostname for multidut environment for all tests
	
other changes per module:
	
	test_chassis.py:
	add tests for new apis for chassis: get_module_index, get_supervisor_slot, get_my_slot, is_modular_chassis

    test_component.py:
	remove range since image_list is list for image in range(image_list) 
	
	test_fan_drawer.py:

	add test for new apis for fan_drawer: get_maximum_consumed_power
	
	test_module.py
	add tests for new apis for module_base: get_description,	get_slot, get_type,	get_oper_status, get_midplane_ip, 
	is_midplane_reachable, get_maximum_consumed_power, reboot, set_admin_state
	
	test_psu.py
	changes changes to run tests on  Sonic chassis,
	add tests for new apis: get_maximum_supplied_power, set_status_master_led, get_status_master_led
	
	
	test_thermal.py
	changes changes to run tests on  Sonic chassis, 
	add tests for new apis for thermal: get_minimum_recorded, get_maximum_recorded
	

#### How did you verify/test it?
         Validated the modified tests against chassis.

### CODE REVIEW

File: tests/common/helpers/platform_api/chassis.py
Comment (jleveque) [inline tests/common/helpers/platform_api/chassis.py:94]: Per [PEP8](https://www.python.org/dev/peps/pep-0008/), there should be two blank lines between top-level function definitions.

File: tests/common/helpers/platform_api/chassis.py
Comment (jleveque) [inline tests/common/helpers/platform_api/chassis.py:182]: Per [PEP8](https://www.python.org/dev/peps/pep-0008/), there should be two blank lines between top-level function definitions.

File: tests/common/helpers/platform_api/fan_drawer.py
Comment (jleveque) [inline tests/common/helpers/platform_api/fan_drawer.py:70]: Per [PEP8](https://www.python.org/dev/peps/pep-0008/), there should be two blank lines between top-level function definitions.

File: tests/common/helpers/platform_api/psu.py
Comment (jleveque) [inline tests/common/helpers/platform_api/psu.py:81]: Per [PEP8](https://www.python.org/dev/peps/pep-0008/), there should be two blank lines between top-level function definitions.

File: tests/common/helpers/platform_api/psu.py
Comment (jleveque) [inline tests/common/helpers/platform_api/psu.py:123]: Per [PEP8](https://www.python.org/dev/peps/pep-0008/), there should be two blank lines between top-level function definitions.

File: tests/platform_tests/api/conftest.py
Comment (jleveque) [inline tests/platform_tests/api/conftest.py:72]: Sugesst breaking this line as follows:
```python
out = duthost.shell('docker exec pmon supervisorctl status platform_api_server',
                    module_ignore_errors=True)[ 'stdout_lines']
```

File: tests/platform_tests/api/test_fan_drawer.py
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:61]: Add space after `=`

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — CHANGES_REQUESTED** [review]
Please fix conflicts.

File: tests/common/helpers/platform_api/module.py
Comment (jleveque) [inline tests/common/helpers/platform_api/module.py:163]: Delete trailing blank lines

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]
Looks good to me. Please wait for another review, as this PR touches a lot of files.

**Comment by rawal01** [comment]
@yxieca @vdahiya12 @sujinmkang  Please review this PR and merge

**Comment by shubav** [comment]
@jleveque, could you help add the second reviewer, whoever you think best knows this area? We rebased and pushed the changes again. please help move this forward. thanks

**Comment by jleveque** [comment]
@yxieca: Can you please review this PR for sonic-mgmt best practices when you have a moment?

File: tests/platform_tests/api/test_chassis_fans.py
Comment (wangxin) [inline tests/platform_tests/api/test_chassis_fans.py:79]: The `platform_api_conn` fixture is parameterized using `enum_rand_one_per_hwsku_hostname`. For such test cases that are also parameterized using `enum_rand_one_per_hwsku_hostname`, I am not sure how pytest will generate the eventual list of parameterized test cases. Have you verified this on a chassis? For example, if there are two nodes (node1, node2), will the generated test cases be something like this:
test_get_name[node1](..., platform_api_conn[node1])
test_get_name[node1](..., platform_api_conn[node2])
test_get_name[node2](..., platform_api_conn[node1])
test_get_name[node2](..., platform_api_conn[node2])

Cases like `test_get_name[node1](..., platform_api_conn[node2])
` and `test_get_name[node2](..., platform_api_conn[node1])` would not make sense. For these cases, test executed on node1 would use platform_api_conn for node2. And test executed on node2 would use platform_api_conn for node1.

If pytest just generate test cases `test_get_name[node1](..., platform_api_conn[node1])` and `test_get_name[node1](..., platform_api_conn[node1])`, then it is not an issue.

Can you please confirm?

Comment (rawal01) [reply tests/platform_tests/api/test_chassis_fans.py:79]: @wangxin I have verified with chassis it is selects same node for test and fixture test_get_name[node1](..., platform_api_conn[node1]), test_get_name[node2](..., platform_api_conn[node2])

Comment (wangxin) [reply tests/platform_tests/api/test_chassis_fans.py:79]: Thanks for confirming this!

File: tests/platform_tests/api/test_chassis_fans.py
Comment (rawal01) [reply tests/platform_tests/api/test_chassis_fans.py:212]: @wangxin test cases you mentioned do not use duthost information like some other require to get facts or compare hence not included here

**Review by wangxin — APPROVED** [review]

**Pull request merged**
